### PR TITLE
Implement message preparation hook

### LINKED ISSUE #142: Ability to hook into ChatViewController/Compose message sending
Is there a recommended way to hook into message sending? Specifically, I'm trying to encrypt a message before sending it over the wire. The combination of final classes (`ChannelPresenter`) and extension functions (`ChatViewController+Composer#createComposeView` and `ChatViewController+Composer#send`) make it hard (impossible?) to hook into sending. 

Here is my source where I'm modifying the `ChatViewController`: https://github.com/psylinse/stream-ios-encrypted-chat/blob/master/ios/EncryptedChat/EncryptedChatViewController.swift

Sorry if I'm missing something obvious!

### PR DESCRIPTION
**Pull request by cardoso**
Implemented this based on ChannelPresenter having other callbacks with similar purpose (altering message properties before sending). But I can take other approaches suggested by review.

It is somewhat similar to the [Android SDK approach](https://github.com/GetStream/stream-chat-android/blob/1f43e1563b835a29f726595e8694099011bbca43/sample/src/main/java/io/getstream/chat/example/CustomMessageInputView2.java#L25)

Also, I opted for making all struct Message's properties variable, to avoid the verbosity & error proneness of requiring to copy all properties. Swift will still enforce immutability by requiring `mutable` in the struct's methods and requiring `var` in external scopes, so I don't see downsides.

Closes #142

### CODE REVIEW

**Comment by cardoso** [comment]
As it is, this PR introduces a bug: when cache is used, urls can't be tapped. Next commit will fix it

**Comment by b-onc** [comment]
@cardoso Thanks a lot for this PR!
After discussing internally, we've decided to implement this in another way, so I'll be closing this PR.
I'll definitely mention you as a contributor for this feature in the CHANGELOG, though!

**Comment by b-onc** [comment]
Hello @cardoso 
After more (and _more_) internal discussions, we've decided this way is the most convenient for the end user. So again, thanks a lot for your PR!
I'll have some comments and then we can merge this in.

File: Sources/Core/Presenter/Channel/ChannelPresenter.swift [hunk 1]
```diff
@@ -23,6 +23,8 @@ public final class ChannelPresenter: Presenter {
     public typealias FileAttachmentExtraDataCallback = (URL, Channel) -> Codable?
     /// A callback type for the adding an extra data for an image attachment.
     public typealias ImageAttachmentExtraDataCallback = (URL?, UIImage?, _ isVideo: Bool, Channel) -> Codable?
+    /// A callback type for preparing the message before sending.
+    public typealias MessageTextPreparationCallback = (Message) -> Message?
```
Comment (b-onc) [inline Sources/Core/Presenter/Channel/ChannelPresenter.swift:27]: Can you name this type `MessagePreparationCallback`?

Comment (cardoso) [reply Sources/Core/Presenter/Channel/ChannelPresenter.swift:27]: Sure!

Comment (cardoso) [inline Sources/Core/Presenter/Channel/ChannelPresenter.swift:27]:
```suggestion
    public typealias MessagePreparationCallback = (Message) -> Message?
```

File: Sources/Core/Presenter/Channel/ChannelPresenter.swift [hunk 2]
```diff
@@ -32,6 +34,8 @@ public final class ChannelPresenter: Presenter {
     public var fileAttachmentExtraDataCallback: FileAttachmentExtraDataCallback?
     /// A callback for the adding an extra data for a file attachment.
     public var imageAttachmentExtraDataCallback: ImageAttachmentExtraDataCallback?
+    /// A callback for preparing the message before sending.
+    public var messagePreparationCallback: MessageTextPreparationCallback?
```
Comment (cardoso) [inline Sources/Core/Presenter/Channel/ChannelPresenter.swift:38]:
```suggestion
    public var messagePreparationCallback: MessagePreparationCallback?
```

File: Sources/Core/Presenter/Channel/ChannelPresenter.swift (hunk 1 above)
Comment (cardoso) [reply Sources/Core/Presenter/Channel/ChannelPresenter.swift]: Done. The first approach I implemented was just changing the text, and I forgot to update the naming.

**Comment by b-onc** [comment]
@cardoso Can you write a CHANGELOG entry for this?

**Comment by psylinse** [comment]
@b-onc any idea when this will go out? Sorry if it's been released and I missed it!

**Comment by b-onc** [comment]
Hey @psylinse we'll release this in 2.1 next week!

**Pull request merged**